### PR TITLE
Role menu refactor

### LINKED ISSUE #117: Build and Destroy Refactor
**Who are you?**
CSE Dev Team Member

**What feature do you want?**
I want to change build and destroy server to build and destroy classes.

**Why do you want this feature?**
Right now, build and destroy rely on a roles_list that is basically a copy of the class list. I want a build and destroy focused on the class channels and roles to eliminate the need for roles_list.

**Additional context**
There is a bigger issue surrounding build and destroy server. Which channels and roles should be persistent? Should there be a build and destroy server that can target persistent channels and roles? Should we add a way to capture the current state of the server (channels and roles currently active)?

### PR DESCRIPTION
**Pull request by NicoleP23**
# Description

Refactor for a `ServerManagement.py` cog using Discord.py 2.0, rebuilt commands `buildserver` into `buildcourses`, `destroyserver` into `destroycourses` and `rolemenu` into `buildrolemenu`. In `buildrolemenu`, instead of using discord components for buttons, we now use discord.py 2.0 for button creation.

## Issues

Closes #117 

## Type of change

Select one or more of the following:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Other (describe below)
   - Refactor of the `ServerManagement.py` cog

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. List any edge cases you tested.
If you come up with any edge cases you didn't test while writing this PR, cancel the PR and test again.

- [ ] Test A
      - ran `-destroycourses`
      - should remove all the courses in the server that were in the csv
- [ ] Test B
      - ran `buildcourses` with an attached csv (csv not required if want to build same ones as last time)
      - should build all the class channels on the csv
- [ ] Test C
      - ran `-buildrolemenu`
      - should build buttons where it's called for the classes

# Checklist:

- [ ] All local commits have been pushed to remote
- [ ] All changes on the base branch have been merged into this branch, either by rebase or merge
- [ ] My code is [PEP-8](https://pep8.org/) compliant (excluding maximum line length, keep that to 100ish characters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added [Google Docstrings](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) for all new functions/methods
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: Cogs/ClassManagement.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        message = '__**CREATE FOLLOWING CATEGORIES**__\n'
+        for category_name in categories:
+            message += f'{category_name}\n'
+        await ctx.send(message)
+
+        # Get confirmation before building channels
+        if not await confirmation(self.bot, ctx, 'build'):
+            return
+        
+        permissions = discord.Permissions(read_messages=True, send_messages=True, embed_links=True,
```
Comment (mbozada) [inline Cogs/ClassManagement.py:130]: This is a reminder that we need to fix this at some point. Issue #137.

File: Cogs/ClassManagement.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+            except FileNotFoundError:
+                pass
+            await ctx.message.attachments[0].save(csv_filepath)
+        
+        # Load roles csv
+        courses_df = pd.read_csv(csv_filepath)
+
+        courses_df = courses_df.dropna(subset=['create_channels'])
+
+        role_names=courses_df["role/link"].to_list()
```
Comment (mbozada) [inline Cogs/ClassManagement.py:115]: I think the "proper" way to iterate like this is to use https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.itertuples.html

I'm not entirely sure, but it will probably look a bit cleaner.

File: Cogs/ClassManagement.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+            courses_df = pd.read_csv(csv_filepath)
+        except FileNotFoundError:
+            raise FileNotFoundError
+
+        # drop cross-listed courses (and other roles if on the file)
+        courses_df = courses_df.dropna(subset=['create_channels'])
+
+        # List of names of categories to be destroyed, as determined by saved csv
+        destroy_category_names = courses_df['text'].tolist()
+        destroy_categories = await self.get_category(ctx, destroy_category_names)
```
Comment (mbozada) [inline Cogs/ClassManagement.py:190]: It might be better to error out if the categories and roles can't be found.

Right now there's some weirdness when using build and destroy.

File: Cogs/ClassManagement.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+        find csv and extracts columns needed through a panda dataframe
+        create the buttons and put them in a view
+        aka the role menu, and send to user
+        """
+
+        # finds csv and extracts appropriate columns using a dataframe
+        csv_filepath = f'role_lists/roles_{ctx.guild.id}.csv'
+        courses_df = pd.read_csv(csv_filepath)
+        short_names = courses_df["text"].to_list()
```
Comment (mbozada) [inline Cogs/ClassManagement.py:239]: Once again, https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.itertuples.html

I'm not sure if it's the best way to do it, but it'll probably look cleaner.

Also, it'd be good to get consistent names for the columns. In fact, once we find some consistent names, we should change the column names in the .csv

**Review by mbozada — CHANGES_REQUESTED** [review]
I like this implementation. It definitely works as needed.

I think we could refine it a little bit, but we're pretty close to being able to delete ServerManagement.

Also, ClassManagement.py should be CourseManagement.py if we want to keep a consistent naming scheme.

**Comment by NicoleP23** [comment]
For `buildrolemenu` buttons go to the proper channel there must be a parameter (a prefix such as CS, CEG, etc.), and the parameter MUST be capitalized.

File: Cogs/ClassManagement.py (hunk 2 above)
Comment (NicoleP23) [reply Cogs/ClassManagement.py]: #151

File: Cogs/ClassManagement.py (hunk 4 above)
Comment (NicoleP23) [reply Cogs/ClassManagement.py]: Issue #151
